src/OhmmsApp/RandomNumberControl.cpp

```diff
@@ -375,4 +373,238 @@ void RandomNumberControl::write(const std::string& fname, Communicate* comm)
+  hout.close();
+}
+
+//Scatter read
+void RandomNumberControl::read_scatter(hdf_archive& hin, Communicate* comm)
+{
+  int nthreads = omp_get_max_threads();
+  std::vector<uint_type> vt, vt_tot, mt, mt_tot;
+  TinyVector<int,3> shape_now(comm->size(), nthreads, Random.state_size()); //current configuration
+  TinyVector<int,2> shape(comm->size()*nthreads, Random.state_size()); //dimensions of children dataset
+  TinyVector<int,3> shape_hdf5(3,0); //configuration when hdf5 file was written
+
+  //grab configuration of threads/procs and Random.state_size() in hdf5 file
+  if(comm->rank() == 0)
+  {
+    hin.push(hdf::main_state);
+    hin.read(shape_hdf5, "nprocs_nthreads_statesize");
+  }
+
+  mpi::bcast(*comm, shape_hdf5);
+
+  //if hdf5 file's configuration and current configuration don't match, abort read
+  if(shape_hdf5[0] != shape_now[0] || shape_hdf5[1] != shape_now[1] || shape_hdf5[2] != shape_now[2])
+  {
+    app_log() << "Mismatched random number generators."
+      << "\n  Number of procs in streams : old=" << shape_hdf5[0] << " new= " << shape_now[0]
+      << "\n  Number of threads in streams : old=" << shape_hdf5[1] << " new= " << shape_now[1]
+      << "\n  State size per stream : old=" << shape_hdf5[2] << " new= " << shape_now[2]
+      << "\n  Using the random streams generated at the initialization.\n";
+    return;
+  }
+  app_log() << "  Restart from the random number streams from the previous configuration.\n";
+
+  vt.resize(nthreads*Random.state_size()); //buffer for random nums in children of each thread
+  mt.resize(Random.state_size()); //buffer for random numbers from single Random object
+
+  if(comm->rank() == 0)
+  {
+    hin.push("random"); //group for children[ip] (Random.object for each thread)
+    vt_tot.resize(nthreads*Random.state_size()*comm->size());
+    hyperslab_proxy<std::vector<uint_type>,2> slab(vt_tot, shape);
+    hin.read(slab,Random.EngineName);
+    hin.pop();
+
+    shape[0] = comm->size(); //reset shape to one thread per process
+    mt_tot.resize(Random.state_size()*comm->size());
+    hin.push("random_master"); //group for single Random object
+    hyperslab_proxy<std::vector<uint_type>,2> slab2(mt_tot, shape);
+    hin.read(slab2,Random.EngineName);
+    hin.close();
+  }
+
+  if(comm->size()>1)
+  {
+    mpi::scatter(*comm,vt_tot,vt); //divide big buffer into on for each proc
+    mpi::scatter(*comm,mt_tot,mt);
+  }
+  else
+  {
+    copy(vt_tot.begin(),vt_tot.end(),vt.begin());
+    copy(mt_tot.begin(),mt_tot.end(),mt.begin());
+  }
+
+  std::vector<uint_type>::iterator vt_it(vt.begin());
+  for(int i=0; i<nthreads; i++, vt_it += shape[1])
+  {
+    std::vector<uint_type> c(vt_it,vt_it+shape[1]);
+    Children[i]->load(c); //read seeds for each thread from buffer back into object
+  }
+  Random.load(mt); //read seeds back into object
+}
+
+//scatter write
+void RandomNumberControl::write_scatter(hdf_archive& hout, Communicate* comm)
```

Comment:
To match the underlying operation, this should be write_gather.
But I wonder if it would be better to use something different than 'scatter/gather' for this name - something that could be the same suffix on both read and write.
The operation is to read or write on MPI rank 0, so maybe 'write_rank_0 / read_rank_0' or 'write_head_node / read_head_node' or something else?